src/TextareaPrompt.php

```diff
@@ -0,0 +1,234 @@
+
+        $fullLines = $currentLines->slice(0, -2);
+
+        $this->cursorPosition = $fullLines->sum() + $newColumn;
+    }
+
+    /**
+     * Handle the down keypress.
+     */
+    protected function handleDownKey(): void
+    {
+        $lines = collect($this->lines());
+
+        // Line length + 1 for the newline character
+        $lineLengths = $lines->map(fn ($line, $index) => mb_strwidth($line) + ($index === $lines->count() - 1 ? 0 : 1));
+
+        $currentLineIndex = $this->currentLineIndex();
+
+        if ($currentLineIndex === $lines->count() - 1) {
+            // They're already at the last line, jump them to the last position
+            $this->cursorPosition = mb_strwidth($lines->implode(PHP_EOL));
+
+            return;
+        }
+
+        // Lines up to and including the current line
+        $currentLines = $lineLengths->slice(0, $currentLineIndex + 1);
+
+        $currentColumn = $currentLines->last() - ($currentLines->sum() - $this->cursorPosition);
+
+        $destinationLineLength = ($lineLengths->get($currentLineIndex + 1) ?? $currentLines->last()) - 1;
+
+        $newColumn = min(max(0, $destinationLineLength), $currentColumn);
+
+        $this->cursorPosition = $currentLines->sum() + $newColumn;
+    }
+
+    /**
+     * The currently visible options.
+     *
+     * @return array<int, string>
+     */
+    public function visible(): array
+    {
+        $this->adjustVisibleWindow();
+
+        $withCursor = $this->valueWithCursor(10_000);
+
+        return array_slice(explode(PHP_EOL, $withCursor), $this->firstVisible, $this->scroll, preserve_keys: true);
+    }
+
+    protected function adjustVisibleWindow(): void
+    {
+        if (count($this->lines()) < $this->scroll) {
+            return;
+        }
+
+        $currentLineIndex = $this->currentLineIndex();
+
+        if ($this->firstVisible + $this->scroll <= $currentLineIndex) {
+            $this->firstVisible++;
+        }
+
+        if ($currentLineIndex === $this->firstVisible - 1) {
+            $this->firstVisible = max(0, $this->firstVisible - 1);
+        }
+
+        // Make sure there are always the scroll amount visible
+        if ($this->firstVisible + $this->scroll > count($this->lines())) {
+            $this->firstVisible = count($this->lines()) - $this->scroll;
+        }
+    }
+
+    /**
+     * Get the index of the current line that the cursor is on.
+     */
+    protected function currentLineIndex(): int
+    {
+        $totalLineLength = 0;
+
+        return (int) collect($this->lines())->search(function ($line) use (&$totalLineLength) {
+            $totalLineLength += mb_strwidth($line) + 1;
+
+            return $totalLineLength > $this->cursorPosition;
+        }) ?: 0;
+    }
+
+    /**
+     * Get the formatted lines of the current value.
+     *
+     * @return array<int, string>
+     */
+    public function lines(): array
+    {
+        // Subtract 2 for the scrollbar
+        $this->maxLineWidth = $this->width - 2;
```

Comment:
Can we subtract 2 from `width` in the renderer and remove the `maxLineWidth` property? The scrollbar feels like a renderer concern. It looks like we're using the current value of `width` in one place in this file, though, and I'm not sure of the consequences of it being reduced by 2.

There is a tiny issue that might be related. If you enter the following:

```
a
bb
ccc
```

With the cursor at the very end, if you press the `Up` arrow, it will move to the beginning of the line instead of going up to the end of the previous line. It only seems to happen when the current line is precisely one character longer than the previous. If it's shorter, equal, or more than one character longer, then it moves up as expected.